File: 22. Generate Parentheses/22. Generate Parentheses.java
import java.util.ArrayList;
import java.util.List;

class Solution {

  public List<String> generateParenthesis(int n) {
    List<String> ans = new ArrayList<>();
    StringBuilder curr = new StringBuilder();

    backtrack(0, 0, n, ans, curr);
    return ans;
  }

  private void backtrack(
    int start,
    int end,
    int n,
    List<String> ans,
    StringBuilder curr
  ) {
    if (curr.length() == 2 * n) {
      ans.add(curr.toString());
      return;
    }

    if (start < n) {
      curr.append("(");
      backtrack(start + 1, end, n, ans, curr);
      curr.deleteCharAt(curr.length() - 1);
    }

    if (end < start) {
      curr.append(")");
      backtrack(start, end + 1, n, ans, curr);
      curr.deleteCharAt(curr.length() - 1);
    }
  }
}

File: 22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        curr = []

        def backtrack(openn, close):
            if openn == close == n:
                ans.append("".join(curr))
                return

            if openn < n:
                curr.append("(")
                backtrack(openn + 1, close)
                curr.pop()

            if close < openn:
                curr.append(")")
                backtrack(openn, close + 1)
                curr.pop()

        backtrack(0, 0)
        return ans
